### ISSUE TITLE
Implement Phoenix security

### ISSUE THREAD

**Issue by bijugs** [bloomberg/chef-bach#1110]
Based this [comment](https://github.com/bloomberg/chef-bach/pull/1079#issuecomment-366024595) 
with HDP 2.6.3, we have acls for phoenix in place.
[PHOENIX-4198|https://issues.apache.org/jira/browse/PHOENIX-4198] Remove need for users to have write permission to SYSTEM.CATALOG table.

To Implement security,

```
cookbooks/bcpc-hadoop/attributes/hbase.rb
+  site_xml['phoenix.schema.isNamespaceMappingEnabled'] = true
+  site_xml['phoenix.schema.mapSystemTablesToNamespace'] = true
+  site_xml['phoenix.acls.enabled'] = true
+  site_xml['hbase.coprocessor.master.classes'] = org.apache.hadoop.hbase.security.access.AccessController
+  site_xml['hbase.coprocessor.region.classes'] = org.apache.hadoop.hbase.security.access.AccessController,org.apache.hadoop.hbase.security.token.TokenProvider,org.apache.hadoop.hbase.security.access.SecureBulkLoadEndpoint
```
Apply Hbase Acl permissions:

```
#To Show all permissions on @SYSTEM NAMESPACE
user_permission '@SYSTEM'

#Grant Read only permission on @SYSTEM Namespace.
grant 'user', 'R' , 'SYSTEM:CATALOG'

#Grant respective 'RWXCA' [READ('R'), WRITE('W'), EXEC('X'), CREATE('C'), ADMIN('A')] permissions on user namespace
grant 'user', 'RWXCA' , '@userNamespace'
```

**Comment by bijugs** [comment]
@saumilmayani... added the Phoenix ``acl`` related properties to ``hbase`` listed above and restarted the VM cluster. Then checked the following sequence of actions and looks like the user ``ubuntu`` is able to delete rows from ``SYSTEM.CATALOG`` even though only 'R' permission is set on the ``SYSTEM`` namespace and ```SYSTEM:CATALOG`` . Did I miss anything? 
```
> grant 'ubuntu', 'R', '@SYSTEM'

User                                          Namespace,Table,Family,Qualifier:Permission
 ubuntu                                       SYSTEM,,,: [Permission: actions=READ]
1 row(s) in 0.6750 seconds

0: jdbc:phoenix:f-bcpc-vm1.bcpc.example.com:2> create table UBUNTU.UBUNTU_TEST3 (id integer not null primary key, name varchar);
No rows affected (2.448 seconds)
0: jdbc:phoenix:f-bcpc-vm1.bcpc.example.com:2> select * from UBUNTU.UBUNTU_TEST3;
+-----+-------+
| ID  | NAME  |
+-----+-------+
+-----+-------+

0: jdbc:phoenix:f-bcpc-vm1.bcpc.example.com:2> upsert into UBUNTU.UBUNTU_TEST3 values(1, 'Val1');
1 row affected (0.016 seconds)
0: jdbc:phoenix:f-bcpc-vm1.bcpc.example.com:2> select * from UBUNTU.UBUNTU_TEST3;
+-----+-------+
| ID  | NAME  |
+-----+-------+
| 1   | Val1  |
+-----+-------+

0: jdbc:phoenix:f-bcpc-vm1.bcpc.example.com:2> delete from SYSTEM.CATALOG where TABLE_NAME = 'UBUNTU_TEST4';
3 rows affected (0.063 seconds)


hbase(main):001:0> grant 'ubuntu', 'R', 'SYSTEM:CATALOG'
0 row(s) in 0.6990 seconds

hbase(main):003:0> user_permission 'SYSTEM:CATALOG'
User                                          Namespace,Table,Family,Qualifier:Permission
 ubuntu                                     SYSTEM,SYSTEM:CATALOG,,: [Permission: actions=READ]
 user                                         SYSTEM,SYSTEM:CATALOG,,: [Permission: actions=READ]

0: jdbc:phoenix:f-bcpc-vm1.bcpc.example.com:2> delete from SYSTEM.CATALOG where TABLE_NAME = 'UBUNTU_TEST3';
3 rows affected (0.049 seconds)
```

**Comment by saumilmayani** [comment]
hi @bijugs is ubuntu also in hbase.super list?

**Comment by bijugs** [comment]
Yikes!! eventhough ``ubuntu`` used didn't have a principal in ``Kerberos``, it is being added to the ``hbase.super`` list. Who knew? Thanks for pointing it out. Will remove it and it should help.

**Comment by bijugs** [comment]
After removing ``ubuntu`` from ``hbase.super`` list Connecting to ``Phoenix`` via ``sqlline.py`` generates this error which shows that ``EXEC`` action is made. Users need ``RX`` on ``SYSTEM:CATALOG``.

```
org.apache.hadoop.hbase.security.AccessDeniedException: org.apache.hadoop.hbase.security.AccessDeniedException: Insufficient permissions (user=ubuntu@BCPC.EXAMPLE.COM, scope=SYSTEM:CATALOG, params=[table=SYSTEM:CATALOG],action=EXEC)
```
Even after providing ``RX`` permissions on ``SYSTEM.CATALOG``
```
hbase(main):006:0> user_permission '@SYSTEM'
User                                                     Namespace,Table,Family,Qualifier:Permission
 ubuntu                                                  SYSTEM,,,: [Permission: actions=READ,EXEC]
2 row(s) in 0.1890 seconds

hbase(main):007:0> user_permission 'SYSTEM:CATALOG'
User                                                     Namespace,Table,Family,Qualifier:Permission
 ubuntu                                                  SYSTEM,SYSTEM:CATALOG,,: [Permission: actions=READ,EXEC]
 ```
``sqlline.py`` throws the following warning but able to act on a ``namespace`` to which it has ``RWXCA`` access

```
Caused by: org.apache.hadoop.hbase.ipc.RemoteWithExtrasException(org.apache.hadoop.hbase.security.AccessDeniedException): org.apache.hadoop.hbase.security.AccessDeniedException: Insufficient permissions (user=ubuntu@BCPC.EXAMPLE.COM, scope=SYSTEM, params=[namespace=SYSTEM,table=SYSTEM:CATALOG],action=CREATE)
```

For e.g. permissions set for ``ubuntu`` user on ``UBUNTU`` namespace
```
hbase(main):019:0> user_permission '@UBUNTU'
User                                                     Namespace,Table,Family,Qualifier:Permission
 ubuntu                                                  UBUNTU,,,: [Permission: actions=READ,WRITE,EXEC,CREATE,ADMIN]
```

**Comment by bijugs** [comment]
Resolved by PR #1110 . Closing.
